EndpointManager and NodeManager Cells

joamaki opened the pull request:
This PR adds cells for EndpointManager and NodeManager (cilium-agent objects snippet): 
```
          Ⓜ️ endpoint-manager (Manages the collection of local endpoints):
              ⚙️ (endpointmanager.EndpointManagerConfig) {
                  EndpointGCInterval: (time.Duration) 5m0s
              }

              🚧 endpointmanager.newDefaultEndpointManager (cell.go:177):
                  ⇨ client.Clientset, endpointmanager.EndpointManagerConfig, hive.Lifecycle
                  ⇦ endpointmanager.EndpointManager, endpointmanager.EndpointsLookup, 
                      endpointmanager.EndpointsModify

          Ⓜ️ node-manager (Manages the collection of Cilium nodes):
              🚧 manager.newAllNodeManager (cell.go:69):
                  ⇨ hive.Lifecycle
                  ⇦ manager.NodeManager
```

The endpoint manager is now exposed as 3 different APIs: EndpointsLookup (for read-only lookups of endpoints), EndpointsModify (for adding and removing endpoints), and EndpointManager (for initialization and operations on all endpoints). This division by functionality adds clarity ("ah this module only needs to look up endpoints") and makes integration testing of cells depending on e.g. endpoint lookup simpler.

Node manager's API is kept as in one chunk as it's relatively small. It is still exposed as an interface rather than as struct pointer, again for improved integration testing and for clarity. Providing interfaces at the cell-level does not go against the "return structs, accept interfaces" rule. That rule is all about loose coupling in public APIs, e.g. library functions should be flexible in what they accept (don't take `*os.File`, but `io.Reader`). Here we're following the same spirit, but at the macro level. With interfaces in the object graph we can choose the implementation to provide to a cell or augment an existing one (cell.Decorate) if needed. This gives maximum flexibility and loose coupling that comes especially handy in integration testing.

joamaki committed:
node: Add cell for node manager

Dropped the unused selector cache and policy triggerer from
node manager and moved the node handler subscription out from
the manager creation. NodeDiscovery no longer needs to Close()
the NodeManager as this is now done by a stop hook in NodeManager.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
endpointmanager: Add endpointmanager.Cell

This moves the creation of the endpoint manager from daemon.go
into pkg/endpointmanager/cell.go.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
hive: Drop left-over debug printf

When printing the lifecycle hooks a left-over debug
print appeared if a start function was not defined
in a hook: "{<nil> 0x223af60} not defined".

Signed-off-by: Jussi Maki <jussi@isovalent.com>

dylandreimerink reviewed (approved):
Very nice, like the separate interfaces. Can't find anything to complain about.

christarazi commented on daemon/cmd/daemon.go:
```diff
@@ -1310,24 +1306,7 @@ func newDaemon(ctx context.Context, cleaner *daemonCleanup,
 
-// WithDefaultEndpointManager creates the default endpoint manager with a
-// functional endpoint synchronizer.
-func WithDefaultEndpointManager(ctx context.Context, clientset client.Clientset, checker endpointmanager.EndpointCheckerFunc) *endpointmanager.EndpointManager {
```
[1]

christarazi commented on pkg/endpointmanager/cell.go:
```diff
@@ -0,0 +1,199 @@
+	Manager EndpointManager
+}
+
+func newDefaultEndpointManager(p endpointManagerParams) endpointManagerOut {
+	checker := endpoint.CheckHealth
+
+	mgr := New(&watchers.EndpointSynchronizer{Clientset: p.Clientset})
+	if p.Config.EndpointGCInterval > 0 {
+		ctx, cancel := context.WithCancel(context.Background())
+		mgr = mgr.WithPeriodicEndpointGC(ctx, checker, p.Config.EndpointGCInterval)
```
`newDefaultEndpointManager` is basically what `WithDefaultEndpointManager` used to be. The latter used to take a root ctx from the Daemon [1], but the former doesn't allow for that. Can we make this code path accept a ctx as to keep the same behavior as prior to this PR?

joamaki replied:
With hive/cells there will be no root context anymore (root ctx was problematic as it provided no mechanism to wait for things to finish, which especially impacts integration testing). Each module is supposed to manage their own context (or similar mechanism) whose lifecycle is determined by the Start and Stop hooks.

christarazi replied:
Ah, thanks for the explanation!

christarazi reviewed (changes requested):
Nice, thanks. One comment below about root context passing. Otherwise LGTM.

joamaki committed:
node: Remove node deletion when stopping and wait for background workers

NodeManager's Stop() was earlier only used by tests so it's behavior was to
clean up datapath state by calling NodeDelete for each node. This is not
the desired behavior as we want to preserve the datapath state when shutting
down in order to not disrupt connectivity.

This cleanup behavior was never used in any tests and prior this was only
triggered by control-plane tests via call to Daemon.Close(), so it's safe
to remove this.

In addition Stop() did not wait for the backgroundSync() goroutine to exit
in Stop(). Fix this by using workerpool.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

tommyp1ckles commented on daemon/cmd/cells.go:
```diff
@@ -52,6 +53,9 @@ var (
 		// observing changes to it.
 		node.LocalNodeStoreCell,
 
+		// EndpointManager maintains a collection of the locally running endpoints.
```
Might be helpful to also note that it actually manages implementing endpoints on the host level (avoiding the word manage, possible...).

joamaki replied:
Hm, I'm not sure that's entirely true, e.g. looking at daemon/cmd/endpoint.go a lot of operations on the endpoint happen directly through the `*Endpoint` object and not via EndpointManager. This comment is essentially the comment that was in `pkg/endpointmanager/manager.go` on `EndpointManager struct`. 

What's definitely clear is that this and other APIs could use a revisit as we modularize these. To not get stuck at the start line I'm trying to keep the API as-is for now.

tommyp1ckles commented on pkg/endpointmanager/cell.go:
```diff
@@ -0,0 +1,199 @@
+)
+
+// Cell provides the EndpointManager which maintains the collection of locally
+// running endpoints (containers). Also exposed are EndpointsLookup and
```
Is this referring to k8s/docker containers?

joamaki replied:
yep, cilium endpoint is a container. https://docs.cilium.io/en/v1.12/concepts/terminology/#endpoints. would you have a suggestion how to better word this?

tommyp1ckles replied:
No I think its ok.

tommyp1ckles commented on pkg/endpointmanager/cell.go:
```diff
@@ -0,0 +1,199 @@
+)
+
+type EndpointsLookup interface {
+	Lookup(id string) (*endpoint.Endpoint, error)
```
nit: no comment

tommyp1ckles commented on pkg/endpointmanager/cell.go:
```diff
@@ -0,0 +1,199 @@
+	// LookupCiliumID looks up endpoint by endpoint ID
+	LookupCiliumID(id uint16) *endpoint.Endpoint
+
+	// LookupContainerID looks up endpoint by Docker ID
```
Docker -> container, k8s isn't necessarily using Docker.

tommyp1ckles commented on pkg/endpointmanager/cell.go:
```diff
@@ -0,0 +1,199 @@
+	LookupIP(ip net.IP) (ep *endpoint.Endpoint)
+
+	// LookupPodName looks up endpoint by namespace + pod name
+	LookupPodName(name string) *endpoint.Endpoint
```
nit: might be helpful to say its the fully qualified name, then provide an example (i.e. ns/pod-0).

tommyp1ckles commented on pkg/endpointmanager/cell.go:
```diff
@@ -0,0 +1,199 @@
+	// GetEndpoints returns a slice of all endpoints present in endpoint manager.
+	GetEndpoints() []*endpoint.Endpoint
+
+	// EndpointExists returns whether the endpoint with id exists.
```
This function seems redundant, could we make the api smaller by just using Lookup (and maybe a helper func)?

tommyp1ckles commented on pkg/endpointmanager/cell.go:
```diff
@@ -0,0 +1,199 @@
+
+type EndpointsModify interface {
+	// AddEndpoint takes the prepared endpoint object and starts managing it.
+	AddEndpoint(owner regeneration.Owner, ep *endpoint.Endpoint, reason string) (err error)
```
nit: bit cleaner to not have the named error return type.

tommyp1ckles commented on pkg/endpointmanager/cell.go:
```diff
@@ -0,0 +1,199 @@
+	AddHostEndpoint(
+		ctx context.Context,
+		owner regeneration.Owner,
+		policyGetter policyRepoGetter,
```
nit: I'm not a fan of this many function args, maybe it's worth using a struct.

tommyp1ckles commented on pkg/endpointmanager/cell.go:
```diff
@@ -0,0 +1,199 @@
+	EndpointResourceSynchronizer
+
+	// InitMetrics hooks the EndpointManager into the metrics subsystem. This can
+	// only be done once, globally, otherwise the metrics library will panic.
```
We can make this a bit safer by implementing this with sync.Once(...) on the init?

tommyp1ckles replied:
Actually, I see that the implementation already does that.

Should this even be part of the interface, it seems implementation specific not  useful outside of init, could we just call InitMetrics on the returned struct from `New(..)`?

joamaki replied:
Agreed. Didn't want to change it yet. I'd punt this to @dylandreimerink metrics refactoring which is coming soon.

dylandreimerink replied:
I don't think you need to worry about the metrics refactoring, that is still some time out, we/I will take care of it then.

tommyp1ckles commented on pkg/endpointmanager/cell.go:
```diff
@@ -0,0 +1,199 @@
+	// have had their PolicyMaps updated against the Endpoint's desired policy state.
+	//
+	// Endpoints will wait on the 'notifyWg' parameter before updating policy maps.
+	UpdatePolicyMaps(ctx context.Context, notifyWg *sync.WaitGroup) *sync.WaitGroup
```
I'm not a huge fan of returning wait groups like this, especially at the interface level. What if we just change this to return a func() type and then have the impl return `wg.Wait`.

tommyp1ckles commented on pkg/endpointmanager/cell.go:
```diff
@@ -0,0 +1,199 @@
+	// list is locked and cannot be modified.
+	// Returns a waiting group that can be used to know when all the endpoints are
+	// regenerated.
+	RegenerateAllEndpoints(regenMetadata *regeneration.ExternalRegenerationMetadata) *sync.WaitGroup
```
Same deal here.

sayboras reviewed (approved):
I assume we need to apply this pattern for other subsystems steadily :heavy_check_mark:
